Sprint 16

medvedevda opened the pull request:
Сделал онбординг и добавил проверку проходил ли пользователь уже его или нет. 
Выполнил задание со звездочкой. Хотел бы получить комментарии по тому как я переписал экраны под архитектуру MVVM. Не получилось реализовать обновление коллекции через performBatch. Сделал через reloadData(). Если дадите какие-нибудь подсказки или наводки - буду благодарен.

Commit:
Добавлен Onboarding, а также проверка его прохождения

Commit:
Переписан экран категорий под архитектуру MVVM. Добавлена анимация после онбординга

Commit:
Переписан экран расписания под архитектуру MVVM. Fix bugs

Commit:
Переписан экран создания нового трекера под архитектуру MVVM.

также исправлена функция сортировки выбранного расписания.

Commit:
Переписан главный экран с трекерами под архитектуру MVVM.

Исправлен баг во время поиска(после нажатия на кнопку отменить появлялась заглушка вместо трекеров)

helecta commented on Tracker/Controllers/Trackers/TrackersViewController.swift:
```diff
@@ -21,24 +20,24 @@ final class TrackerViewController: UIViewController, TrackerViewControllerProtoc
         setupNavigationController()
         addTarget()
         checkCellsCount(image: Resourses.Images.trackerEmptyImage!, text: "Что будем отслеживать?")
```
Нужно стараться избегать использовать force операции(тут  force unwrapping). Т.к они могут приводить к падению приложения, считается что падение приложения хуже, чем не рабочее состояние.
Тут можно переписать без него, например если вдруг нет ресурса можно заполнить пустой картинкой: UIImage()

helecta commented on Tracker/Controllers/Trackers/TrackersViewController.swift:
```diff
@@ -106,76 +107,73 @@ final class TrackerViewController: UIViewController, TrackerViewControllerProtoc
-              let presenter = presenter else { return UICollectionViewCell() }
+        trackersViewModel.fetchCompletedCategoriesFromStore()
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "TrackerCollectionViewCell", for: indexPath) as? TrackersCollectionViewCell else { return UICollectionViewCell() }
+        let days = trackersViewModel.getCompletedCategories()
```
Название не соответствует значению. Почему это дни?

helecta commented on Tracker/Controllers/Trackers/TrackersViewController.swift:
```diff
@@ -106,76 +107,73 @@ final class TrackerViewController: UIViewController, TrackerViewControllerProtoc
     
     func numberOfSections(in collectionView: UICollectionView) -> Int {
-        return presenter?.getVisibleCategories().count ?? 0
+        return trackersViewModel.visibleTrackers.count
```
Почему бы во вью модели сразу не возвращать visibleCategories ?

helecta commented on Tracker/Controllers/Trackers/TrackersViewController.swift:
```diff
@@ -51,32 +50,32 @@ final class TrackerViewController: UIViewController, TrackerViewControllerProtoc
+        trackersView.trackersCollectionView.reloadData()
     }
     
     func checkCellsCount(image: UIImage, text: String) {
```
Метод не удовлетворяет принципам SOLID. 
Из названия не понятно, что он делает. Кажется что должен был бы возвращать количество ячеек, но он ничего не возвращает. Внутри у него идет разветвление, которое логически не очень связано. Это видно даже из параметров которые передаются, они используются только в первом случае.

Дальше, внутри этого метода у тебя зашита бизнес логика: Если количество трекеров равно 0, то надо показать одно состояние экрана, если нет то другое.
Так вот бизнес логика должна быть спрятана во вью модели, а в контроллер передаваться уже само состояние.

Я бы сделала, чтобы вью контроллер отслеживал изменение visibleTrackers(либо категорий, если ты решишь переделать на visibleCategories). Как только изменяется количество, то обновляется состояние экрана.(У тебя уже даже есть это отслеживание в bindViewModel)

В контроллере создаем `func reloadViews()`
Дальше тебе надо понять, показывать заглушки или же коллекции. 
Т.е запрашиваем эту информацию у вью модели: `trackersViewModel.isEmptyState()`
Если состояние пустое: то запрашиваем данные для заполнения заглушки:
```
trackerView.emptyImage.image = trackersViewModel.emptyImage
trackerView.emptyLabel.text = trackersViewModel.emptyText
```
Причем какую картинку вернуть и какой текст определяется внутри вьюмодели(в зависимости от того из поиска или из выбора даты пришли)

Если не пустое, то все тоже самое что у тебя сейчас.


Еще можно добавить паттерн state machine. Подробнее если интересно, можешь почитать [тут](https://medium.com/swlh/state-machine-with-viewmodel-7c1eac3b6fb9)

helecta replied:
Тогда и методы setupTrackersFromDatePicker и setupTrackersFromTextField будут не нужны
Тебе нужно будет только сделать, что при изменении даты меняется дата во вью модели и это запускает обновление `visibleTrackers`, и аналогично для поиска.

helecta replied:
Основная идея что вью контроллеры мы стараемся делать максимально тупыми, они отвечают как экран выглядит. И постоянно спрашивает у вью модели, какие данные показать: какой текст, какую картинку, сколько ячеек и т.п.
Изменил пользователь текст поиска, вью контроллер передал это изменение во вью модель, вью модель что-то внутри себя сделала и вернула какие данные теперь надо показать, сколько теперь ячеек, или может вообще заглушку, если да то какую

helecta commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/CategoryViewModel.swift:
```diff
@@ -0,0 +1,37 @@
+        visibleCategories = dataProvider.updateCategoryViewModel()
+    }
+    
+    func getSelectedCategory() -> String {
```
Можно сделать короче:

```
var selectedCategory: String {
        get {
            dataProvider.selectedCategory ?? ""
        }
        
        set {
            dataProvider.selectedCategory = newValue
        }
    }
```

Тогда эти два метода не нужны

helecta commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/CategoryViewController.swift:
```diff
@@ -62,31 +72,27 @@ final class CategoryViewController: UIViewController, CategoryViewControllerProt
+        cell.accessoryType = cell.label.text == categoryViewModel.getSelectedCategory() ? .checkmark : .none
         
-        if indexPath.row + 1 == dataProvider.numberOfCategories {
+        if indexPath.row + 1 == categoryViewModel.numberOfCategories {
```
Лучше сделать во вью модели метод `categoryViewModel.isLastCategory(at: indexPath)`
Тогда код сразу будет удобно и понятно читать, и опять таки бизнес логика определения критериев последней страницы будет убрана во вью модель

helecta commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/CategoryViewController.swift:
```diff
@@ -40,7 +50,7 @@ final class CategoryViewController: UIViewController, CategoryViewControllerProt
     }
     
     func checkCellsCount() {
```
комментарий к этому методу аналогичен комменту к [TrackersViewController.swift](https://github.com/D1mon227/Tracker/pull/4/files#diff-309be61b952db94bd6929ac58aa168318041f7e15bb8df591fa3a8535bccfab1)

helecta replied:
Тут меньше бизнес логики, поэтому не так очевидно. Так что опишу и тут каким хотелось бы видеть вью контроллер.
Метод `checkCellsCount` переписать так `reloadViews`:
```
func reloadViews() {
        if categoryViewModel.isEmpty() {
            reloadEmptyState()
        } else {
            reloadCategoryTableView()
        }
    }

private func reloadEmptyState() {
        view.addSubview(categoryView.emptyImage)
        view.addSubview(categoryView.emptyLabel)
        categoryView.categoryTableView.removeFromSuperview()

        categoryView.emptyImage.snp.makeConstraints { make in
            make.center.equalToSuperview()
        }

        categoryView.emptyLabel.snp.makeConstraints { make in
            make.centerX.equalToSuperview()
            make.leading.trailing.equalToSuperview().inset(16)
            make.top.equalTo(categoryView.emptyImage.snp.bottom).offset(8)
        }
    }
    
private func reloadCategoryTableView() {
        categoryView.emptyImage.removeFromSuperview()
        categoryView.emptyLabel.removeFromSuperview()
        view.addSubview(categoryView.categoryTableView)

        categoryView.categoryTableView.snp.makeConstraints { make in
            make.top.equalTo(categoryView.categoryLabel.snp.bottom).offset(38)
            make.leading.trailing.equalToSuperview().inset(16)
            make.bottom.equalTo(categoryView.addCategoryButton.snp.top).offset(-16)
        }
        categoryView.categoryTableView.reloadData()
    }

private func bindViewModel() {
        categoryViewModel.$visibleCategories.bind { [weak self] _ in
            guard let self = self else { return }
            self.reloadViews()
            
        }
    }
```

helecta replied:
Может показаться что только нейминг изменился, но это не совсем так. Если ты теперь прочитаешь код, то там нет бизнес логики, только изменения вью в зависимости от состояния.
Если состояние пустое, то вид экрана один, если состояние не пустое, то другой.

helecta replied:
По поводу анимации, можно было бы сделать так:
```
        let oldCount = categoryView.categoryTableView.numberOfRows(inSection: 0)
        let newCount = categoryViewModel.numberOfCategories

        categoryView.categoryTableView.performBatchUpdates {
            var indexPaths: [IndexPath] = []
            for i in oldCount..<newCount {
                indexPaths.append(IndexPath(row: i, section: 0))
            }
            categoryView.categoryTableView.insertRows(at: indexPaths, with: .automatic)
        } completion: { _ in }
```

Вместо `categoryView.categoryTableView.reloadData()`, но из-за дизайна ячеек тогда нужно не только `insertRows` делать, но и `reloadRows` для предыдущей ячейки, поэтому спокойно можешь тут использовать `reloadData()`. Но если очень хочется, то смотреть надо в эту сторону.

Получилось бы так, нужно было бы каждый раз обновлять предыдущую ячейку
```
categoryView.categoryTableView.performBatchUpdates {
            var insertIndexPaths: [IndexPath] = []
            var reloadIndexPaths: [IndexPath] = []
            for i in oldCount..<newCount {
                insertIndexPaths.append(IndexPath(row: i, section: 0))
            }

            let prevIndexRow = newCount-2
            if oldCount != newCount &&  prevIndexRow > 0  {
                reloadIndexPaths.append(IndexPath(row: prevIndexRow, section: 0))
            }

            categoryView.categoryTableView.reloadRows(at: reloadIndexPaths, with: .automatic)
            categoryView.categoryTableView.insertRows(at: insertIndexPaths, with: .automatic)
        } completion: { _ in }
```

helecta commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/CategoryViewModel.swift:
```diff
@@ -0,0 +1,37 @@
+        getVisibleCategories()
+    }
+    
+    var numberOfCategories: Int? {
```
Лучше здесь убрать опциональность, и добавить значение по умолчанию

Commit:
Исправлены ошибки после ревью:

-вынесена логика из categoryVC во viewModel
-вынесена логика из trackersVC во viewModel
-исправлен нейминг в trackersVC
-удалена конструкция force unwrapping

Commit:
Исправлен Onboarding.

Оставил только один viewController, в котором меняю backgroundView и label.
